File: Infra/Infra/Face.cpp
#include "Face.hpp"

CppClassNew(Face)

Int Face_Init(Int o)
{
    Face* m;
    m = CP(o);

    QString familyString;
    Int ua;
    ua = CastInt(&familyString);

    String_QStringSet(ua, m->Family);

    int sizeU;
    sizeU = m->Size;

    int weightU;
    weightU = m->Weight;

    bool italicU;
    italicU = m->Italic;

    bool underlineU;
    underlineU = m->Underline;

    bool overlineU;
    overlineU = m->Overline;

    bool strikeoutU;
    strikeoutU = m->Strikeout;

    m->Intern = new QFont(familyString, sizeU, weightU, italicU);
    m->Intern->setStyleHint(QFont::AnyStyle, QFont::PreferAntialias);
    m->Intern->setUnderline(underlineU);
    m->Intern->setOverline(overlineU);
    m->Intern->setStrikeOut(strikeoutU);
    return true;
}

Int Face_Final(Int o)
{
    Face* m;
    m = CP(o);

    delete m->Intern;
    return true;
}

CppField(Face, Family)
CppField(Face, Size)
CppField(Face, Weight)
CppField(Face, Italic)
CppField(Face, Underline)
CppField(Face, Overline)
CppField(Face, Strikeout)

Int Face_Intern(Int o)
{
    Face* m;
    m = CP(o);
    Int a;
    a = CastInt(m->Intern);
    return a;
}

File: Avalon/Avalon.Draw/Face.cs
namespace Avalon.Draw;

public class Face : Any
{
    public override bool Init()
    {
        base.Init();
        this.InternIntern = InternIntern.This;
        this.InternInfra = InternInfra.This;

        this.InternName = this.InternInfra.StringCreate(this.Name.Value);

        ulong sizeU;
        sizeU = (ulong)(this.Size);
        ulong weightU;
        weightU = (ulong)(this.Weight);
        ulong italicU;
        italicU = (ulong)(this.Italic ? 1 : 0);
        ulong underlineU;
        underlineU = (ulong)(this.Underline ? 1 : 0);
        ulong overlineU;
        overlineU = (ulong)(this.Overline ? 1 : 0);
        ulong strikeoutU;
        strikeoutU = (ulong)(this.Strikeout ? 1 : 0);

        this.Intern = Extern.Face_New();
        // Extern.Face_NameSet(this.Intern, this.InternName);
        Extern.Face_SizeSet(this.Intern, sizeU);
        Extern.Face_WeightSet(this.Intern, weightU);
        Extern.Face_ItalicSet(this.Intern, italicU);
        Extern.Face_UnderlineSet(this.Intern, underlineU);
        Extern.Face_OverlineSet(this.Intern, overlineU);
        Extern.Face_StrikeoutSet(this.Intern, strikeoutU);
        Extern.Face_Init(this.Intern);
        return true;
    }

    public virtual bool Final()
    {
        Extern.Face_Final(this.Intern);
        Extern.Face_Delete(this.Intern);

        this.InternInfra.StringDelete(this.InternName);
        return true;
    }

    public virtual String Name { get; set; }
    public virtual long Size { get; set; }
    public virtual long Weight { get; set; }
    public virtual bool Italic { get; set; }
    public virtual bool Underline { get; set; }
    public virtual bool Overline { get; set; }
    public virtual bool Strikeout { get; set; }

    private InternIntern InternIntern { get; set; }
    private InternInfra InternInfra { get; set; }
    internal virtual ulong Intern { get; set; }
    private ulong InternName { get; set; }
}